samza-core/src/main/java/org/apache/samza/processor/StreamProcessor.java

```diff
@@ -153,54 +153,68 @@ public State getState() {
+   * {@link JobCoordinator} is given for this {@link StreamProcessor}.
+   * @param config configuration required to launch {@link JobCoordinator} and {@link SamzaContainer}
    * @param customMetricsReporters metric Reporter
-   * @param streamTaskFactory task factory to instantiate the Task
-   * @param processorListener  listener to the StreamProcessor life cycle
+   * @param taskFactory task factory to instantiate the Task
+   * @param processorListener listener to the StreamProcessor life cycle
+   * @param jobCoordinator the instance of {@link JobCoordinator}
    */
-  public StreamProcessor(Config config, Map<String, MetricsReporter> customMetricsReporters,
-                         StreamTaskFactory streamTaskFactory, StreamProcessorLifecycleListener processorListener) {
-    this(config, customMetricsReporters, streamTaskFactory, processorListener, null);
-  }
-
-  /* package private */
-  private JobCoordinator getJobCoordinator() {
-    String jobCoordinatorFactoryClassName = new JobCoordinatorConfig(config).getJobCoordinatorFactoryClassName();
-    return Util.getObj(jobCoordinatorFactoryClassName, JobCoordinatorFactory.class).getJobCoordinator(config);
-  }
-
-  @VisibleForTesting
-  JobCoordinator getCurrentJobCoordinator() {
-    return jobCoordinator;
+  public StreamProcessor(Config config, Map<String, MetricsReporter> customMetricsReporters, TaskFactory taskFactory,
+                  ProcessorLifecycleListener processorListener, JobCoordinator jobCoordinator) {
+    Preconditions.checkNotNull(processorListener, "ProcessorListener cannot be null.");
+    this.taskFactory = taskFactory;
+    this.config = config;
+    this.taskShutdownMs = new TaskConfigJava(config).getShutdownMs();
+    this.processorListener = processorListener;
+    this.customMetricsReporter = customMetricsReporters;
+    this.jobCoordinator = (jobCoordinator != null) ? jobCoordinator : createJobCoordinator();
+    this.jobCoordinatorListener = createJobCoordinatorListener();
+    this.jobCoordinator.setListener(jobCoordinatorListener);
+    ThreadFactory threadFactory = new ThreadFactoryBuilder().setNameFormat(CONTAINER_THREAD_NAME_FORMAT).setDaemon(true).build();
+    this.executorService = Executors.newSingleThreadExecutor(threadFactory);
+    // TODO: remove the dependency on jobCoordinator for processorId after fixing SAMZA-1835
+    this.processorId = this.jobCoordinator.getProcessorId();
   }
 
-  StreamProcessor(Config config, Map<String, MetricsReporter> customMetricsReporters, Object taskFactory,
-                  StreamProcessorLifecycleListener processorListener, JobCoordinator jobCoordinator) {
-    Preconditions.checkNotNull(processorListener, "ProcessorListener cannot be null.");
+  /**
+   * Same as {@link #StreamProcessor(Config, Map, TaskFactory, ProcessorLifecycleListener, JobCoordinator)}, except
+   * there is a {@link StreamProcessorListenerFactory} as input instead of {@link ProcessorLifecycleListener}.
```

Comment:
The reason that I did not fold them together is the fact that I want to call factory.createInstance() after the StreamProcessor's other internal states are already setup, in case the factory.createInstance() method depends on an internal state variable in StreamProcessor has already been initialized. I can instead, making the non factory constructor as a caller to the factor constructor by providing a lambda as the factory.